executor/src/witgen/processor.rs

```diff
@@ -148,27 +148,42 @@ impl<'a, 'b, T: FieldElement, Q: QueryCallback<T>, CalldataAvailable>
     ) -> Result<Constraints<&'a PolynomialReference, T>, EvalError<T>> {
         let mut outer_assignments = vec![];
 
-        while let Some(SequenceStep {
-            row_delta,
-            identity,
-        }) = sequence_iterator.next()
-        {
+        while let Some(SequenceStep { row_delta, action }) = sequence_iterator.next() {
             let row_index = (1 + row_delta) as usize;
-            let progress = match identity {
-                IdentityInSequence::Internal(identity_index) => {
+            let progress = match action {
+                Action::InternalIdentity(identity_index) => {
                     self.process_identity(row_index, identity_index)?
                 }
-                IdentityInSequence::OuterQuery => {
+                Action::OuterQuery => {
                     let (progress, new_outer_assignments) = self.process_outer_query(row_index)?;
                     outer_assignments.extend(new_outer_assignments);
                     progress
                 }
+                Action::Queries => self.process_queries(row_index),
             };
             sequence_iterator.report_progress(progress);
         }
         Ok(outer_assignments)
     }
 
+    fn process_queries(&mut self, row_index: usize) -> bool {
+        let global_row_index = self.row_offset + row_index as u64;
+        let row_pair = RowPair::new(
+            &self.data[row_index],
+            &self.data[row_index + 1],
+            global_row_index,
+            self.fixed_data,
+            UnknownStrategy::Unknown,
+        );
```

Comment:
I think this should be extracted into a function, if I see it correctly, we have 3 copies of this code now. Alse the `self.row_offset + row_index` calculation can be done there.

Reply:
Oh yeah, I know that problem, but I fear there is no solution.